### PR TITLE
grub: Safely join paths in the grub parser

### PR DESCRIPTION
**Pull request by rjoleary**
This prevents paths from escaping the mount point such as
`filepath.Join("/tmp/mount-sda1/", "../../etc/passwd")`.

Signed-off-by: Ryan O'Leary <ryanoleary@google.com>

### CODE REVIEW

**Review by rminnich — APPROVED** [review]
we're making copies of this function everywhere. Go upstream dumped on my proposal to put such a function in filepath. Should we put this in uroot/util?

If you're ok with having it here that's fine oto.

**Comment by rjoleary** [comment]
I see. This should be merged with #1817

**Review by GanShun — APPROVED** [review]

File: pkg/upath/safejoin.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import (
+	"fmt"
+	"path/filepath"
+)
+
+// SafeFilepathJoin safely joins two paths path1+path2. The resulting path will
+// always be contained within path1 even if path2 tries to escape with "../".
+//
+// Note: This only works on Unix-like file systems.
```
Comment (rminnich) [inline pkg/upath/safejoin.go:15]: it would work on plan9 too but nobody calls that unix-like :-)
why would this not work on windows?

**Review by rminnich — APPROVED** [review]

File: pkg/upath/safejoin.go (hunk 1 above)
Comment (rjoleary) [reply pkg/upath/safejoin.go:15]: On windows, they have paths like `C:\Documents\`, so prefixing with a `/` does not make sense.

Comment (rminnich) [reply pkg/upath/safejoin.go:15]: so should it have a build tag for ! windows?

**Comment by rjoleary** [comment]
Re-implementing symlinks in user-space is a bit complicated. Let's look a the chroot approach.

**Pull request closed without merging**